enhancements/insights/config-api.md

```diff
@@ -0,0 +1,245 @@
+This would also allow to expose the basic Insights analysis results in the corresponding status section. This functionality now relies on the in-cluster Prometheus (and it will probably not be fully available in the HyperShift architecture).
+Finally, one of the reasons for this enhancement is to define a new way for a user to trigger the Insights data collection.  
+
+### Goals
+
+The goal of this enhancement is to propose a basic structure for a new Insights configuration API and define new configuration options. This includes following:
+
+- Introduce the basic configuration options in the API
+- Define a new configuration option providing a way to trigger on demand Insights gathering
+- Define a new configuration option providing a way to include/exclude particular Insights data gatherers
+- Provide Insights status & reporting in the status field in the new API
+
+### Non-Goals
+
+- Move the majority of the current configuration options to a new configuration API - Insights has quite a lot of configuration options
+  and it does not make sense to expose everything in the OpenShift API 
+
+## Proposal
+
+The proposal is to introduce a new Insights `CustomResourceDefinition` in the OpenShift API providing a way to configure the Insights operator and to report its status. The new CRD should follow a common CRD structure providing `spec` and `status` attributes. 
+
+```go
+type InsightsSpec struct {
+	// Gathering spec attribute includes all the configuration options related to
+	// gathering of the Insights archive and its uploading to the ingress.
+	GatheringConfig GatheringConfig `json:"gatheringConfig,omitempty"`
+}
+
+type InsightsStatus struct {
+	// GatheringStatus provides basic information about the last Insights gathering.
+	GatheringStatus GatheringStatus `json:"gatheringStatus,omitempty"`
+	// SimpleContentAccessStatus provides basic information about the SCA controller.
+	SimpleContentAccessStatus SimpleContentAccessStatus `json:"simpleContentAccessStatus,omitempty"`
+	// ClusterTransferStatus provides basic information about the ClusterTransfer controller.
+	ClusterTransferStatus ClusterTransferStatus `json:"clusterTransferStatus,omitempty"`
+	// ReportStatus provides general Insights analysis results.
+	ReportStatus ReportStatus `json:"reportStatus,omitempty"`
+}
+
+type GatheringConfig struct {
+	// EnableObfuscation is flag enabling the global obfuscation of the IP addresses and base domain
+	// in the Insights archive data.
+	EnableObfuscation bool `json:"enableObfuscation,omitempty"`
+	// Enables user to force Insights data gathering by setting a new reason. Gathering can be forced even
+	// when all the gatherers are disabled.
```

Comment:
what actually gets gathered if all the gatherers are disabled?

Reply:
Nothing. There's no new archive -> nothing to upload. The gathering cycle is skipped and/or is basically turned off. I can add it here.

Reply:
yeah i think it's worth clarifying, lest people think there is some baseline/default gathering that happens even when other gatherers are disabled.

Reply:
Can we say:

> Gathering can be forced even when all the gatherers are disabled, although the operator will not attempt to upload the resulting empty archive.

or something?  Because currently you just seem to be saying "this won't cause any errors" without spelling out "because it's a no-op".

Reply:
@wking Hmmm I don't think that's the desirable behavior here. It doesn't make sense to gather empty archive in this case IMO. My thinking was that when the gathering was forced then everything was gathered. In other words `ForceGatherReason` overrides the `DisabledGatherers` attribute. It was discussed here https://github.com/openshift/enhancements/pull/1037#discussion_r812405104

The use case I see is that you're in a disabled (or even disconnected) env but you want to gather "manually" something from time to time.

Reply:
> My thinking was that when the gathering was forced then everything was gathered.

Ah, that was not clear to me.  I was seeing this as just a manual trigger of whatever polled gather you'd configured.  Because folks doing that manual gather might still have privacy concerns if the dislike some.particular gatherer, right?

Reply:
@tremes when you answered me you said nothing would be gathered if all gatherers were disabled.

now you're saying that force will always gather everything regardless of which gatherers are enabled?

As a user I would definitely not expect that.  I would expect force to just be "gather right now".  If you want to gather something that isn't configured, then you should change the config and then force the gather.

If my cluster is on its knees and i need to force gather for support, i want to be able to tightly control what gets gathered in case some gatherers aren't going to be able to function in the current state of the cluster.  

Either way, the godoc for the force field should very explicitly state what the behavior will be, with respect to the Gatherers value, but again i feel pretty strongly that force should respect the configured gatherers setting

Reply:
I wasn't thinking about combining of the forced gathering and deactivating some gatherers properly or I was little confused (sorry). I mean I was thinking mainly about this scenario:

- Insights operator is disabled (disabledGatherers: ["all"])
- user wants to get one-off data so the gathering is forced (and he/she might or might not upload it manually)

It seemed to me easier to gather everything in this case then updating the `disabledGatherers` first and then forcing the gathering. I think you're right that respecting the configured gatherers is more coherent. So in the case of `disabledGatherers: ["all"]` and forced gathering, nothing happens right? We can say (in the `GatheringStatus` probably) something like that gathering is disabled, please enable some of the gatherers and force it again.

Reply:
> I think you're right that respecting the configured gatherers is more coherent.

it also makes the api more powerful (in that users can fully control what gets gathered even in the case of a "forced" gather) rather than taking away control from users, so i think it's better for both reasons.

it is true that for someone who has disabled all gatherers, they have to do slightly more configuration to do the forced gather, but it's still the same resource they are editing so not hugely painful (they will also then want to re-disable the gatherers after the force)

> So in the case of disabledGatherers: ["all"] and forced gathering, nothing happens right?

right

Reply:
Yes. Thanks. This was an important discussion. I updated the text as well as the comment.